### DISCUSSION TITLE
Welcome to spotify-qt Discussions!

### DISCUSSION THREAD

**Discussion by kraxarn** [kraxarn/spotify-qt#37] [General]
## Welcome to Discussions
Leave your thoughts and feelings below if you want to.

Note: Don't report bugs or issues here, please open up an issue instead as it's easier to track.

**Comment by J-Cake** [comment]
I **really really** like this. It's Qt so its *f*****g* fast and cause its Qt, it fits with the KDE theme beautifully. I only have one small thing I'd like to mention: The media keys (DBus I believe) don't work at all. There seems to be no way to control the media at all other than from the app. I can't use the `playerctl <pause | play etc.>` commands which is a bit of a nuisance, but other than that, it's brilliant

**Reply by kraxarn** [reply]
Glad to hear you're liking it ☺️

The media controller has always been very buggy, see #4, specially under kde for some reason, as gnome for example seems to work fine from my testing. `playerctl` should work fine though, did you compile it yourself or use the snap? If you're using the snap, it might be related to #38. When the next update comes out, and you're still experiencing issues with using `playerctl`, please comment on the issue, or open a new one. Also make sure "media controller" is enabled in the application settings.

**Reply by J-Cake** [reply]
Thanks for your reply, I've enabled the *media controller* option to no avail. I have no control over the playback outside of the app. There is no output at all

**Reply by kraxarn** [reply]
Did you try again in v3.3? It includes some media controller improvements, more specifically when using the snap package.

**Comment by zihenlin** [comment]
Hey, coming from spotify-tui, spotify-qt gives much better experience. The dark theme is superb despite I'm not using KDE. It fits in XFCE pretty well. Thank you for the amazing effort :)

**Comment by jsiulian** [comment]
Great idea, I was so glad to see a lightweight player! I do seem to have one problem though, it won't accept my path to spotifyd, I am getting the `Failed to apply setting "spotifyd path". Check your settings and try again.` message. 
Looking at the code it seems to get to `applyFail("spotifyd path");` which if I understand correctly thinks my client path field is empty. It definitely isn't though, I have "/snap/spotifyd/current/bin/spotifyd" in there. Reading your comment above, about bugs and KDE, I do want to mention I am on Kubuntu 20.10, I hope that's not the issue and I'm just being silly. Can you think of a cause for this?

Also, something else minor I have noticed, SpotifyQt seems to look for the `spotifyd.conf` in `/etc/spotifyd/`, while the spotifyd official page mentions it needs to be in `/etc/xdg/spotifyd/` . Not a big deal, just thought I'd mention it

Many thanks!

**Reply by kraxarn** [reply]
It's only the media player functionality that's a bit buggy under kde, the application itself should work fine, as i primarily use kde myself. Did you also install the application itself from the snap? In that case, there might be some permission issue.

Those were just the used paths when spotifyd support was first implemented, but I'll be sure to update the paths then, thanks.

**Reply by jsiulian** [reply]
Yes that's right, both installed through snap; will try installing using something else in that case. Appreciate the help, thanks!

**Reply by kraxarn** [reply]
Probably a snap permission issue in that case, and probably not something that can be fixed, but a note should be added in the app about it.

**Reply by jsiulian** [reply]
Understood, will confirm tonight when I get to my kubuntu box

**Reply by jsiulian** [reply]
Ok so I sort of solved the issue - it was due to the fact that I was starting SpotifyQt from Snap (because I didn't have a shortcut after installation). Once I created a shortcut on the desktop it worked (and even the initial client secret dialog looked different!). I am not an expert in linux :)

**Reply by kraxarn** [reply]
Ah, I'm honestly not too familiar with how snap works, and the snap version of the application is not regularly tested, but then i should probably look into creating a shortcut with the snap.

**Reply by J-Cake** [reply]
@jsiulian I noticed this as well. When I started the app from the command line, `/snap/spotifyqt/spotifyqt` I got the themeless app. I just had to make a shortcut to it on the desktop and it worked normally. I have everything installed through snap and it works fine. One thing to note, though, is the snap version of the app doesn't have clean media key integration on KDE, so I had to remap the media keys to the playerctl commands. Not a big issue, but worth noting nonetheless.

**Reply by jsiulian** [reply]
I basically found out that Snap is a thing by reading this thread, I don't understand much of what happens. 

@J-Cake In that same folder `/snap/spotifyqt/` I noticed there is another file, can't remember the exact name but it's a *.sh launcher - I haven't tried it though. Is it supposed to do something different?

**Reply by kraxarn** [reply]
That's what i mean about buggy media controller support under kde, it's not exclusive for the snap. I also remapped the media keys, and kind of forgot about it, but it's definitely something that needs to be fixed.

Again, I'm not too familiar with snaps, but it does include a desktop launcher script, otherwise qt applications won't work properly. I think you're also supposed to start the application using `snap run spotify-qt`, rather than starting the executable directly, so that might just run the script.

**Reply by J-Cake** [reply]
@jsiulian I don't see such a file, but I don't seem to need it. Take a look at this. This is my launcher config that I can place anywhere and it just works.

![capture_2021-01-28_14:18:38](https://user-images.githubusercontent.com/29561775/106085117-b8883680-6173-11eb-99fc-5934001525fb.png)
![capture_2021-01-28_14:17:47](https://user-images.githubusercontent.com/29561775/106085047-9abad180-6173-11eb-9d89-33219a188983.png)

The shortcut is stored in `~/.local/share/applications/spotify-qt.desktop` and executes `/snap/spotify-qt/current/bin/spotify-qt` when it's opened. This allows you to find the file through the KDE menus and uses the system theme. It also allows you to map the media keys to it as `playerctl` can now find it.

**Reply by jsiulian** [reply]
Sorry guys, been busy for a few days. 
@kraxarn is seems that launching it using `snap run spotify-qt` launches it in unskinned mode

@J-Cake that's exactly what I did as well

Here's a screenshot of 2 instances launched differently, guess which is which!

![image](https://user-images.githubusercontent.com/2658641/106367240-fda99580-6338-11eb-8edf-f121e7a98108.png)

I understand even less about snap than I thought I did now lol!

I did notice a bug, in that double clicking any song will only play the current song (can even see it is sending the same command in spotifyd log) - I will open a bug for that

**Reply by jsiulian** [reply]
I do want to add that I love the app and I am gonna continue using it - maybe even on windows.

I wonder if adding a "Now playing" view is planned? It would be nice to see what list of songs it is playing through

**Reply by J-Cake** [reply]
The *Now Playing* idea is a good one, I meant to mention it, but I never did. As for why the app remains unskinned, I can't say I can help you other than to try to create a shortcut to the executable and launch that under KDE, because that works for me. Otherwise, I don't know enough about how the system works to be much help.

**Reply by kraxarn** [reply]
If you mean some sort of view where you can see all the upcoming tracks, it's actually a limitation with the Spotify API and therefore can't be easily implemented. The app itself could try to figure out the upcoming tracks, but as soon as you enable shuffle, it's basically impossible.

Because snaps run in their own little container, they don't use the system theme. It is possible to mess around with it enough to make it use the gtk theme, but i just didn't bother. If you launch the executable directly, it probably launches it outside of the container and is then allowed to use the system theme.

**Reply by J-Cake** [reply]
Just relating to your comment on what's currently playing; can't you just look that up from spotifyd or have I terribly misunderstood what spotifyd does?
Thanks

**Comment by toadative** [comment]
Is there an easy way to install this for Windows? Currently I'm trying to build from source because all of the packages are for Linux. I got cmake and qt5 installed but now I'm getting this error and I'm way over my head:
> CMake Error at CMakeLists.txt:33 (find_package):
  By not providing "FindQt5.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Qt5", but
  CMake did not find one.

>  Could not find a package configuration file provided by "Qt5" with any of
  the following names:

>    Qt5Config.cmake
    qt5-config.cmake

>  Add the installation prefix of "Qt5" to CMAKE_PREFIX_PATH or set "Qt5_DIR"
  to a directory containing one of the above files.  If "Qt5" provides a
  separate development package or SDK, be sure it has been installed.

Do you know how to solve this, or if there is an easier way to do this?

**Reply by kraxarn** [reply]
It's regularly tested on Windows, so it does build and run just fine. I personally have a msys2 installation, which makes everything a lot easier, but it looks like you just need to tell cmake where qt is installed for it to work, for example by using `cmake -DQt5_DIR="..."`, depending on how you installed it.

I have considered providing windows binaries for a while, but partly because of issues like these, and, most importantly, the incomplete and buggy windows support with librespot and spotifyd, i didn't really think it made much sense. That's why I wanted to wait until i properly ported a playback client into the app before i started providing windows binaries, but because it has been requested multiple times now, i might look into it anyway.

**Reply by J-Cake** [reply]
Could you set up some sort of Github Actions build chain? cause I'd be interested in Windows binaries as well ;P

**Reply by kraxarn** [reply]
Automated Windows builds are now available built as a GitHub action! From v3.6 and onwards, they will be attached to the release, but for builds from each commit, you can download it by going to "Actions > MSYS2 > The latest successful run > spotify-qt under Artifacts" and it should download a portable zip file. Note that you need to be signed it for it to show up for some reason.

I don't currently have Windows installed, but I tried it in a Windows 8.1 virtual machine, and it seems to work fine. I'm also quite unfamiliar with Windows build systems, so it uses MinGW, which results in a quite large file size, but you shouldn't need to install anything else, like some VC++ runtime. Please try it out and let me know how it works :smile:

**Comment by ciompi** [comment]
Hi kraxarn I enjoy your work, the qt interface shows my playlists and is ok
but I can't start the spotifid daemon as seems that within the program I don't have the rights to read or access to any folder, so neither the one containing spotifyd
may be I have done some mistakes? I have installed daemon and qt interface via snap on ubuntu software app
I am running a lenovo atom netbook with xubuntu 18.05
any suggestion?
grazie
gianmario nava

**Reply by ciompi** [reply]
btw qt-spotyfy correctly detect my android app on my phone and took control of it
great!
gmn

**Reply by kraxarn** [reply]
Yeah, I'm aware of the permission issues with the snap, although I'm unable to reproduce any issues with it myself. Snap does have an API spotify-qt could make use of, and due to the amount of issues reported related to this, it might be worth looking into.

**Reply by ciompi** [reply]
Thanks for the quick reply, I will try to compile myself. big achievement for me...
grazie e arrivederci
gmn
